packages/storage-driver-cloudinary/src/index.ts

```diff
@@ -0,0 +1,387 @@
+		this.apiKey = config.apiKey;
+		this.apiSecret = config.apiSecret;
+		this.cloudName = config.cloudName;
+		this.accessMode = config.accessMode;
+	}
+
+	private fullPath(filepath: string) {
+		return normalizePath(join(this.root, filepath), { removeLeading: true });
+	}
+
+	private toFormUrlEncoded(obj: Record<string, string>, options?: { sort: boolean }) {
+		let entries = Object.entries(obj);
+
+		if (options?.sort) {
+			entries = entries.sort(([keyA], [keyB]) => keyA.localeCompare(keyB));
+		}
+
+		return new URLSearchParams(entries).toString();
+	}
+
+	/**
+	 * Generate the Cloudinary SHA1 signature for the given payload
+	 * @see https://cloudinary.com/documentation/signatures
+	 */
+	private getFullSignature(payload: Record<string, string>) {
+		const denylist = ['file', 'cloud_name', 'resource_type', 'api_key'];
+
+		const signaturePayload = Object.fromEntries(
+			Object.entries(payload).filter(([key]) => denylist.includes(key) === false)
+		);
+
+		const signaturePayloadString = this.toFormUrlEncoded(signaturePayload, { sort: true });
+
+		return createHash('sha1')
+			.update(signaturePayloadString + this.apiSecret)
+			.digest('hex');
+	}
+
+	/**
+	 * Creates inline URL signature for use with the image reading API
+	 * @see https://cloudinary.com/documentation/advanced_url_delivery_options#generating_delivery_url_signatures
+	 */
+	private getParameterSignature(filepath: string) {
+		return `s--${createHash('sha1')
+			.update(filepath + this.apiSecret)
+			.digest('base64url')
+			.substring(0, 8)}--`;
+	}
+
+	private getTimestamp() {
+		return String(new Date().getTime());
+	}
+
+	/**
+	 * Used to guess what resource type is appropriate for a given filepath
+	 * @see https://cloudinary.com/documentation/image_transformations#image_upload_note
+	 */
+	private getResourceType(filepath: string) {
+		const fileExtension = extname(filepath);
+		if (IMAGE_EXTENSIONS.includes(fileExtension)) return 'image';
+		if (VIDEO_EXTENSIONS.includes(fileExtension)) return 'video';
+		return 'raw';
+	}
+
+	/**
+	 * For Cloudinary Admin APIs, the file extension needs to be omitted for images and videos. Raw
+	 * on the other hand requires the extension to be present.
+	 */
+	private getPublicId(filepath: string) {
+		const resourceType = this.getResourceType(filepath);
+		if (resourceType === 'raw') return filepath;
+		return parse(filepath).name;
+	}
+
+	/**
+	 * Generates the Authorization header value for Cloudinary's basic auth endpoints
+	 */
+	private getBasicAuth() {
+		const creds = `${this.apiKey}:${this.apiSecret}`;
+		const base64 = Buffer.from(creds).toString('base64');
+		return `Basic ${base64}`;
+	}
+
+	async read(filepath: string, range?: Range) {
+		const resourceType = this.getResourceType(filepath);
+		const fullPath = this.fullPath(filepath);
+		const signature = this.getParameterSignature(fullPath);
+		const url = `https://res.cloudinary.com/${this.cloudName}/${resourceType}/upload/${signature}/${fullPath}`;
```

Comment:
> However I do notice that there is a possibility that this can also be https://mycloud-res.cloudinary.com instead of https://res.cloudinary.com/mycloud, for private CDN distributions that is available for advanced plans or higher as mentioned in their docs: https://cloudinary.com/documentation/solution_overview#urls_and_endpoints
> 
> Not sure how edge-case-y that is, but maybe we can turn this into a private method or conditional statement to form the correct URL (maybe modifiable via cloudinary-driver-specific environment variable)

I think I'm gonna wait on this until the need arises. I don't know how common it is to have those custom slugs, and similarly to the sha-1 thing, it seems to be an enterprise enterprise only thing you have to reach out to support for in order to get "enabled" 🤔 Lets keep it simple for now, should be easy enough to add in later 👍🏻